### ISSUE TITLE
small implementation error in DeepFool attack

### ISSUE THREAD

**Issue by LaurenzBeck** [Harry24k/adversarial-attacks-pytorch#51]
I am quite new to adverserial attacks, so the current behaviour might be intended.

I currently use your DeepFool implementation as an estimate for the distance to the decision boundary of a classifier, as suggested in the paper: https://arxiv.org/abs/2002.01810v1

If the model already predicts a wrong label for a given input sample, then the perturbation returned by the DeepFool attack is the original image (https://github.com/Harry24k/adversarial-attacks-pytorch/blob/528ab8074794cf9952e58815617e64b5c73e2046/torchattacks/attacks/deepfool.py#L74), whereas I think that just returning `torch.zeros_like(image)` would be formally correct.

Changing this would not only correct the estimate on the decision boundary, but also the doubling of the final adversarial sample (image + perturbation) if the classifier misclassified from the start.

**Comment by LaurenzBeck** [comment]
My suggestion might be missleading, I meant returning (True, pre, torch.zeros_like(image) in https://github.com/Harry24k/adversarial-attacks-pytorch/blob/528ab8074794cf9952e58815617e64b5c73e2046/torchattacks/attacks/deepfool.py#L74

**Comment by LaurenzBeck** [comment]
I just tested my suggestion and noticed that it has the unwanted side-effect of also returning zeros when we achieve a successful misclassification through the adverserial attack.

I solved this by adding `curr_steps` to the call to `self._forward_indiv` in https://github.com/Harry24k/adversarial-attacks-pytorch/blob/528ab8074794cf9952e58815617e64b5c73e2046/torchattacks/attacks/deepfool.py#L55

in the `self._forward_indiv` we then have to add a check for `curr_steps` to be zero:
```python
def _forward_indiv(self, image, label, curr_steps):
        image.requires_grad = True
        fs = self.model(image)[0]
        _, pre = torch.max(fs, dim=0)
        if pre != label:
            if curr_steps == 0:
                return (True, pre, torch.zeros_like(image))
            else:
                return (True, pre, image)
```

**Comment by Harry24k** [comment]
Thanks for the suggestion. However, a function `_forward_indiv()` has to output the adversarial image rather than the perturbation for all cases, because `forward()` replace the final adv_image with `_forward_indiv()` as follows:
 https://github.com/Harry24k/adversarial-attacks-pytorch/blob/528ab8074794cf9952e58815617e64b5c73e2046/torchattacks/attacks/deepfool.py#L56.

(When the prediction is already the wrong label.)
https://github.com/Harry24k/adversarial-attacks-pytorch/blob/528ab8074794cf9952e58815617e64b5c73e2046/torchattacks/attacks/deepfool.py#L99

(When the prediction becomes the wrong label.)
https://github.com/Harry24k/adversarial-attacks-pytorch/blob/528ab8074794cf9952e58815617e64b5c73e2046/torchattacks/attacks/deepfool.py#L99

Thus, when `_forward_indiv()` outputs `torch.zeros_like(image)`, the final adv_image is replaced with `torch.zeros_like(image)` instead of the original images, which induces incorrect adv_images.

For the case that the perturbation is needed instead of the final adv_image, you can code like `perturbation = adv_images - images`. If I misunderstood, please leave a comment :)

**Comment by LaurenzBeck** [comment]
That seams reasonable and consistent with the documented use:
```python
Examples::
        >>> attack = torchattacks.DeepFool(model, steps=50, overshoot=0.02)
        >>> adv_images = attack(images, labels) #-> adv_images not adv_perturbations or adv_noise
```

I was testing this Attacker on my dataset by plotting:
    1) original image with target and prediction in title
    2) adv_image
    3) original image + adv_image with adv_prediction in title

It seamed to me, that adv_images only contained the adverserial perturbations, because the signals were quite weak (to be expected from DeepFool) and did not contain a sample-specific mean shift (special to the dataset I was testing on).
So I thought that `original image + adv_image` had to be the adv_image and adv_image was really just the perturbation.

The only problem was then, that the original images were present in the adv_images when the classifier predicted wrong in the first case (hence this issue here).

I might be making a fool of myself, but I think that in either intended case (adv_images with or without original image) adjustments have to be made.

**Comment by LaurenzBeck** [comment]
Waiiit.
I just saw, that I overlooked the condition, that the images have to be in range [0, 1).
The images I tested on were in range (-1, 1). So the clamping at 0 had mislead a lot of my asumptions:
https://github.com/Harry24k/adversarial-attacks-pytorch/blob/528ab8074794cf9952e58815617e64b5c73e2046/torchattacks/attacks/deepfool.py#L98

Today I will dig into the paper to see where this constraint comes from.
With this new insight I think that the implementation is most probably correct.
Sorry for the inconvenience!
